### PR TITLE
Unify version pattern, allow local dandiset URLs

### PR DESCRIPTION
**Pull request by AlmightyYakob**
When doing work in the API, I noticed that the regex pattern for the published dandiset URL was hardcoded to `https://dandiarchive.org`. I thought this should be prod/dev agnostic, as the other URL patterns are.

Also, I noticed there was some duplication regarding version patterns, so I unified them.

### CODE REVIEW

**Comment by satra** [comment]
@jwodder - can you please confirm that the CLI test failures are known issues at this point in time ?

**Comment by jwodder** [comment]
@satra They are not.  I see messages like `'https://dandiarchive.org/dandiset/000002/0.220304.1811' does not match '^http://localhost(:\\\\d+)?/dandiset/\\\\d{6}/\\\\d+\\\\.\\\\d+\\\\.\\\\d+$'`, which are clearly caused by this PR.

@AlmightyYakob If `DANDI_ALLOW_LOCALHOST_URLS` is set, then `DANDI_ARCHIVE_PATTERN` should allow localhost *in addition* to https://dandiarchive.org.

File: dandischema/models.py [hunk 1]
```diff
@@ -42,17 +42,21 @@
 # Local or test deployments of dandi-api will insert URLs into the schema that refer to the domain
 # localhost, which is not a valid TLD. To make the metadata valid in those contexts, setting this
 # environment variable will use a less restrictive pydantic field that allows localhost.
+DANDI_ARCHIVE_PATTERN = r"https://dandiarchive.org"
 if "DANDI_ALLOW_LOCALHOST_URLS" in os.environ:
     HttpUrl = AnyHttpUrl  # noqa: F811
+    DANDI_ARCHIVE_PATTERN = r"(https://dandiarchive.org)|(http://localhost)(:\d+)?"
```
Comment (jwodder) [inline dandischema/models.py:48]:
```suggestion
    DANDI_ARCHIVE_PATTERN = r"(https://dandiarchive.org|http://localhost(:\d+)?)"
```

**Review by jwodder — CHANGES_REQUESTED** [review]

**Comment by AlmightyYakob** [comment]
@jwodder are these failing codecov tests a problem?

**Comment by jwodder** [comment]
@AlmightyYakob No.

**Pull request merged**